File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
def solution(x, y, n):
    answer = 0
    dp = [-1 for _ in range(y+1)]
    dp[x] = 0
    
    for i in range(x, y+1):
        if dp[i] == -1:
            continue
            
        if i+n <= y:
            dp[i+n] = min(dp[i]+1, dp[i+n]) if dp[i+n] != -1 else dp[i]+1
        if i*2 <= y:
            dp[i*2] = min(dp[i]+1, dp[i*2]) if dp[i*2] != -1 else dp[i]+1
        if i*3 <= y:
            dp[i*3] = min(dp[i]+1, dp[i*3]) if dp[i*3] != -1 else dp[i]+1
    
    return dp[y]

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        int answer = 0;
        
        int[] dp = new int[y+1];
        Arrays.fill(dp, -1);
        dp[x] = 0;
        
        while (x <= y){
            if(dp[x] == -1) {
                ++x;
                continue;
            }
            
            if(x+n <= y){
                dp[x+n] = dp[x+n] != -1 ? Math.min(dp[x]+1, dp[x+n]) : dp[x]+1;
            }
            if(x*2 <= y){
                dp[x*2] = dp[x*2] != -1 ? Math.min(dp[x]+1, dp[x*2]) : dp[x]+1;
            }
            if(x*3 <= y){
                dp[x*3] = dp[x*3] != -1 ? Math.min(dp[x]+1, dp[x*3]) : dp[x]+1;
            }
            x++;
        }
        
        // BFS
//         Queue<Integer> queue = new ArrayDeque<>();
//         queue.add(x);

//         while (!queue.isEmpty()) {
//             int cur = queue.poll();
//             // 이미 y에 도달한 경우 더 이상 진행X
//             if (cur == y) return dp[cur];
//             for (int next : new int[]{cur+n, cur*2, cur*3}) {
//                 if (next > y) continue;
//                 if (dp[next] == -1 || dp[next] > dp[cur] + 1) {
//                     dp[next] = dp[cur]+1;
//                     queue.add(next);
//                 }
//             }
//         }
//         return -1;
        
        return dp[y];
    }
}